Fix SIOF in torch using caffe2 registry

bwasti opened the pull request:
Summary:
This resolves the issues associated with caffe2 initialization (specifically the REGISTER_FUNCTION_SCHEMA_OPERATOR calls) being run after Torch's static op registration calls.

The fix employs a meyer's singleton wrapped by the constructor of a type.  Everything is placed inside a macro to make it easier for users to use.

Differential Revision: D13854306

smessmer reviewed (approved):
please fix CI before landing

bwasti committed:
bleh

Differential Revision: D13882416

fbshipit-source-id: 0670dd75dbd5ad4c66900ff59b827ddaec1274bd

bwasti committed:
Fix SIOF in torch using caffe2 registry (#16473)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/16473

This resolves the issues associated with caffe2 initialization (specifically the REGISTER_FUNCTION_SCHEMA_OPERATOR calls) being run after Torch's static op registration calls.

The fix employs a meyer's singleton wrapped by the constructor of a type.  Everything is placed inside a macro to make it easier for users to use.

Reviewed By: smessmer

Differential Revision: D13854306

fbshipit-source-id: c0298f6757f004af69acc2c4d5d03b8b4267d3ad
